modules/appnexusAstBidAdapter.js

```diff
@@ -26,357 +24,323 @@ const NATIVE_MAPPING = {
-          }
-        }
+    const memberIdBid = bidRequests.find(hasMemberId);
+    const member = memberIdBid ? parseInt(memberIdBid.params.member, 10) : 0;
 
-        if (bid.mediaType === 'video') { tag.require_asset_url = true; }
-        if (bid.params.video) {
-          tag.video = {};
-          // place any valid video params on the tag
-          Object.keys(bid.params.video)
-            .filter(param => VIDEO_TARGETING.includes(param))
-            .forEach(param => tag.video[param] = bid.params.video[param]);
-        }
+    const payload = {tags: [...tags], user: userObj};
+    if (member > 0) {
+      payload.member_id = member;
+    }
+    const payloadString = JSON.stringify(payload);
+    return {
+      type: POST,
+      endpoint: ENDPOINT,
+      data: payloadString,
+    };
+  },
 
-        if (bid.params.user) {
-          userObj = {};
-          Object.keys(bid.params.user)
-            .filter(param => USER_PARAMS.includes(param))
-            .forEach(param => userObj[param] = bid.params.user[param]);
+  /**
+   * Unpack the response from the server into a list of bids.
+   *
+   * @param {*} serverResponse A successful response from the server.
+   * @return {Bid[]} An array of bids which were nested inside the server.
+   */
+  interpretResponse: function(serverResponse) {
+    const parsed = JSON.parse(serverResponse);
+    const bids = [];
+    parsed.tags.forEach(serverBid => {
+      const rtbBid = getRtbBid(serverBid);
+      if (rtbBid) {
+        if (rtbBid.cpm !== 0 && SUPPORTED_AD_TYPES.includes(rtbBid.ad_type)) {
+          const bid = newBid(serverBid, rtbBid);
+          bid.mediaType = parseMediaType(rtbBid);
+          bids.push(bid);
         }
-
-        return tag;
-      });
-
-    if (!utils.isEmpty(tags)) {
-      const payloadJson = {tags: [...tags], user: userObj};
-      if (member > 0) {
-        payloadJson.member_id = member;
       }
-      const payload = JSON.stringify(payloadJson);
-      ajax(ENDPOINT, handleResponse, payload, {
-        contentType: 'text/plain',
-        withCredentials: true
-      });
-    }
-  };
-
-  /* Notify Prebid of bid responses so bids can get in the auction */
-  function handleResponse(response) {
-    let parsed;
-
-    try {
-      parsed = JSON.parse(response);
-    } catch (error) {
-      utils.logError(error);
-    }
+    });
+    return bids;
+  }
+}
 
-    if (!parsed || parsed.error) {
-      let errorMessage = `in response for ${baseAdapter.getBidderCode()} adapter`;
-      if (parsed && parsed.error) { errorMessage += `: ${parsed.error}`; }
-      utils.logError(errorMessage);
-
-      // signal this response is complete
-      Object.keys(bidRequests)
-        .map(bidId => bidRequests[bidId].placementCode)
-        .forEach(placementCode => {
-          bidmanager.addBidResponse(placementCode, createBid(STATUS.NO_BID));
-        });
-      return;
+// TODO: Patch UserSync into the spec, once the UserSync PR has been merged.
```

Comment:
that's a good question... and I don't know. I actually chatted with our team about it, but didn't walk away with any super clear answers.

I'm not familiar enough with AdTech to know how it's done... but I could imagine three places where user syncs _might_ come from.

1. Constant values. "Always ping this usersync endpoint on every auction"
2. Server-side logic: "Always ping this endpoint to _get_ the usersync info. It will return one or more syncs"
3. Winning bids. "Because this bid made it into the auction, drop this usersync pixel"

Depending on which cases are worth supporting, I'd choose different APIs. It'll require adding one or more optional properties to the spec:

For (1), it'd be `getSyncs()`
For (2), it'd be `getSyncs(done)`
For (3), it'd be `getSyncs(serverResponse)`

In all cases, the return value (or callback argument) would just be an object with the data which we need to call `registerSync` in #1549.

Do you have any better background or opinion on it? I'm interested to hear what other adapter maintainers think too, if they see this.